### DISCUSSION TITLE
Control cookies

### DISCUSSION THREAD

**Discussion by rmccar** [ONSdigital/design-system#715] [Feedback]
Use this place to discuss the [control cookies pattern](https://ons-design-system.netlify.com//patterns/cookie-settings/) in the [ONS Design System](https://ons-design-system.netlify.com)

**Comment by jrbarnes9** [comment]
Cookie banner buttons spacing lost on overlapping 500px breakpoint
![Screenshot 2020-07-09 at 13 51 17](https://user-images.githubusercontent.com/43346934/87042746-fe87a800-c1eb-11ea-8341-056e045b0ec7.png)

**Comment by jrbarnes9** [comment]
Saving the cookies form in the DS example sends you to the previous page you were on

**Reply by jrbarnes9** [reply]
Raised #1899

**Comment by jrbarnes9** [comment]
Cookie banner: Params missing from pattern macro options tables

- `secondaryButtonUrl`

**Reply by jrbarnes9** [reply]
Raised #1912

**Comment by jrbarnes9** [comment]
We need to review our cookie banner to ensure it's compliant with the regulations as many gov services have recently updated their banner to make sure the user can equally turn on or turn off all non-essential cookies in one click.

**Reply by jrbarnes9** [reply]
Raised #1913

**Comment by jrbarnes9** [comment]
DAC: Move skip link after cookie banner 

Would also be useful to:
- Add Cookie banner, phase banner etc in the blocks exploded view to users can see what they should be used for
- Add a page example to the skip link component docs to see where it's positioned, how it links to main etc.

**Reply by jrbarnes9** [reply]
#2337

**Comment by Paul-Joel** [comment]
Bug: Revised cookie banner code in 54.0.1 errors when used in Jinja if no params are provided. Params are not being tested they exist before trying to use them, a common issue between nunjucks and jinja code

**Comment by mr-nick17** [comment]
I would like to understand the rationale of creating a [valid JSON object](https://github.com/ONSdigital/design-system/blob/2027b02f283fbc0b3775296a249b8cfb34817229/src/js/cookies-functions.js#L1) then converting it to a [valid JSON string](https://github.com/ONSdigital/design-system/blob/2027b02f283fbc0b3775296a249b8cfb34817229/src/js/cookies-functions.js#L45) (all 👍 good stuff so far) to then immediately invalidate the JSON object by replacing the double quotes with a single quote which is not valid JSON. As a consuming downstream service, I would expect to work with valid recognisable objects. Deviation from such should be clearly documented with reasoning/commentary provided.

**Reply by MagdalenaLarge** [reply]
Hi @mr-nick17 , thank you for raising this issue up. Team will need to investigate into it, so I will raise a ticket for them and keep you updated about the progress. In case it is very urgent issue - please let me know and we will prioritise accordingly. Thanks.

**Reply by SriHV** [reply]
Hi @mr-nick17, the reason for using single quotes was to make it compatible with Census Respondent Home, a public facing landing page where members of the public launched their Census EQ in 2021. I spoke to few teams and they told me it was now decommissioned. We will raise a new card to look into using double quotes for the cookies JSON but I am not sure on the timescales for this as it will be a low priority and has the potential to impact many services that use the Design System so we would need to do extensive testing with them before making the change.

**Reply by mr-nick17** [reply]
Thanks for the update. Going forward I would recommend that the design-system avoids customisations for specific systems. The downstream service is responsible for implementing overrides. Otherwise you end up in this position. However, I am curious to understand the 'impact on many services' you are trying to mitigate and whom you consult with. This change could be part of the 2 major api increments the service manual has committed to. I am not aware of a consultation process the service-manual goes through that determines what is included within the major increments. If this consultation exists, please ensure all teams within Dissemination are included.

**Reply by rmccar** [reply]
This is something we do avoid, but in the past (especially around Census) the DS was much more tightly coupled with the services in SDC. This is because the DS was initially created for this purpose so there may be some instances where things like this are still in there.
The 'impact on many services' we mentioned would be the potential impact of making this change on the services especially in SDC that already use this component and who we consulted with on this would be the teams that work on those services themselves.
Are you referring to the 2 major releases a year we have planned to make? If this is the case, yes we plan to make breaking changes like this in these releases but doing the work on this issue is still quite low priority for us compared with other pieces of work we have.

**Reply by mr-nick17** [reply]
The 2 major releases was indeed referring to the ones that you have planned to make and my thinking was that as they are major releases, there does not need to be backward compatibility (nor is it expected) and the 'impact on many services' does not require such consideration other that some decent release notes and comms. However, I was unaware that such tight coupling still existed. What are the timelines to remove this coupling? 

I appreciate that the work is 'low priority' and arguably relatively easy. My concern is that bugs, however small, should not be allowed to manifest and as more services consume the design-system, more will require a patch and more teams will be affected. Take the public facing website for instance - at least 3 separate teams have already been required to write overrides to read the malformed JSON, all of these teams will be required to update their apps when a fix comes along as this bug has manifested this far. Not forgetting the teams who have already consumed this and written their own overrides. Good practise is to address bugs as soon as they are identified. As more time elapses the domain knowledge of where these fixes will be required will leak and the risk of losing our analytics increases. 

Some places which have had to override the malformed JSON:
- https://github.com/ONSdigital/dp-renderer/blob/0dd8cfa1d6292405009fe86e18c70a62f719c074/assets/templates/partials/gtm-data-layer.tmpl#L3 
- https://github.com/ONSdigital/dis-wagtail/blob/86c997f476cc6eea51a40d41d1fd5b0f26e44e92/cms/jinja2/templates/base.html#L103
- https://github.com/ONSdigital/dp-census-atlas/blob/9ad5345c1bcdf3263cadf93db27caa30a3a1a4f6/src/components/OnsAnalyticsBanner.svelte#L23

**Reply by rmccar** [reply]
I understand your concerns but like I said this work is lower priority than other work we have at the moment so this will have to go on our backlog for the time being

**Reply by mr-nick17** [reply]
Please can you send me the ticket so I can track/watch it?

**Reply by rmccar** [reply]
Here is the ticket: https://jira.ons.gov.uk/browse/ONSDESYS-473
